apis/Google.Cloud.Spanner.V1/Google.Cloud.Spanner.Data/ClientPool.cs

```diff
@@ -0,0 +1,100 @@
+﻿using System;
+using System.Collections.Concurrent;
+using System.Threading;
+using System.Threading.Tasks;
+using Google.Api.Gax.Grpc;
+using Google.Apis.Auth.OAuth2;
+using Google.Cloud.Spanner.V1;
+using Google.Cloud.Spanner.V1.Logging;
+using Grpc.Auth;
+
+namespace Google.Cloud.Spanner
+{
+    internal static class ClientPool
+    {
+        private static readonly ConcurrentDictionary<ClientPoolKey, ClientPoolEntry> s_clientEntryPool =
+            new ConcurrentDictionary<ClientPoolKey, ClientPoolEntry>();
+
+        public static int Timeout { get; set; }
+
+        public static Task<SpannerClient> AcquireClientAsync()
+        {
+            return AcquireClientAsync(null, null);
+        }
+
+        public static async Task<SpannerClient> AcquireClientAsync(ITokenAccess credentials, ServiceEndpoint endpoint)
+        {
+            var key = new ClientPoolKey {
+                Credential = credentials ?? ApplicationDefault.Instance,
+                Endpoint = endpoint ?? SpannerClient.DefaultEndpoint
+            };
+            var poolEntry = s_clientEntryPool.GetOrAdd(key, k => new ClientPoolEntry(key));
+            var result = await poolEntry.AcquireClientFromEntryAsync().ConfigureAwait(false);
+            Logger.LogPerformanceCounter("SpannerClient.Count", () => s_clientEntryPool.Count);
+            return result;
+        }
+
+        /// <summary>
+        ///     Callers should be careful not to call AcquireClient while CloseAll is still running.
+        ///     It is possible to obtain a stale client.
```

Comment:
It really cannot lock this down because it'd be a race condition from the caller.  For example if you call Acquire + closeAll on different threads, there is simply no way for CloseAll to know that its shutting down a "used" channel.  It cannot tell if a client created 5 milliseconds ago and returned is ok to shut down or not.  All it can do it shut down everything in use now.  This is what the grpc layer does and seems good enough here.